File: app/src/main/java/com/zwstudio/logicgamesandroid/slitherlink/domain/SlitherLinkGameState.java
package com.zwstudio.logicgamesandroid.slitherlink.domain;

import com.zwstudio.logicgamesandroid.logicgames.domain.CellsGameState;
import com.zwstudio.logicgamesandroid.logicgames.domain.Graph;
import com.zwstudio.logicgamesandroid.logicgames.domain.LogicGamesHintState;
import com.zwstudio.logicgamesandroid.logicgames.domain.Node;
import com.zwstudio.logicgamesandroid.logicgames.domain.Position;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.F;

import static fj.data.Array.arrayArray;
import static fj.data.List.iterableList;

/**
 * Created by zwvista on 2016/09/29.
 */

public class SlitherLinkGameState extends CellsGameState<SlitherLinkGame> {
    public SlitherLinkObject[][] objArray;
    public Map<Position, LogicGamesHintState> pos2state = new HashMap<>();

    public SlitherLinkGameState(SlitherLinkGame game) {
        super(game);
        objArray = new SlitherLinkObject[rows() * cols()][];
        for (int i = 0; i < objArray.length; i++) {
            objArray[i] = new SlitherLinkObject[4];
            Arrays.fill(objArray[i], SlitherLinkObject.Empty);
        }
    }

    public SlitherLinkObject[] get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public SlitherLinkObject[] get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, SlitherLinkObject[] dotObj) {
        objArray[row * cols() + col] = dotObj;
    }
    public void set(Position p, SlitherLinkObject[] obj) {
        set(p.row, p.col, obj);
    }

    private void updateIsSolved() {
        isSolved = true;
        for (Map.Entry<Position, Integer> entry : game.pos2hint.entrySet()) {
            Position p = entry.getKey();
            int n2 = entry.getValue();
            int n1 = 0;
            if (get(p)[1] == SlitherLinkObject.Line) n1++;
            if (get(p)[2] == SlitherLinkObject.Line) n1++;
            if (get(p.add(new Position(1, 1)))[0] == SlitherLinkObject.Line) n1++;
            if (get(p.add(new Position(1, 1)))[3] == SlitherLinkObject.Line) n1++;
            pos2state.put(p, n1 < n2 ? LogicGamesHintState.Normal : n1 == n2 ? LogicGamesHintState.Complete : LogicGamesHintState.Error);
            if (n1 != n2) isSolved = false;
        }
        if (!isSolved) return;
        Graph g = new Graph();
        Map<Position, Node> pos2Node = new HashMap<>();
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                int n = arrayArray(get(p)).filter(o -> o == SlitherLinkObject.Line).length();
                switch (n) {
                case 0:
                    continue;
                case 2:
                    {
                        Node node = new Node(p.toString());
                        g.addNode(node);
                        pos2Node.put(p, node);
                    }
                    break;
                default:
                    isSolved = false;
                    return;
                }
            }

        for (Position p : pos2Node.keySet()) {
            SlitherLinkObject[] dotObj = get(p);
            for (int i = 0; i < 4; i++) {
                if (dotObj[i] != SlitherLinkObject.Line) continue;
                Position p2 = p.add(SlitherLinkGame.offset[i]);
                g.connectNode(pos2Node.get(p), pos2Node.get(p2));
            }
        }
        g.setRootNode(iterableList(pos2Node.values()).head());
        List<Node> nodeList = g.bfs();
        int n1 = nodeList.size();
        int n2 = pos2Node.values().size();
        if (n1 != n2) isSolved = false;
    }

    public boolean setObject(SlitherLinkGameMove move) {
        Position p1 = move.p;
        boolean isH = move.objOrientation == SlitherLinkObjectOrientation.Horizontal;
        int i1 = isH ? 1 : 2;
        SlitherLinkObject o = get(p1)[i1];
        if (o.equals(move.obj)) return false;
        Position p2 = p1.add(SlitherLinkGame.offset[isH ? 1 : 2]);
        int i2 = isH ? 3 : 0;
        get(p2)[i2] = get(p1)[i1] = move.obj;
        updateIsSolved();
        return true;
    }

    public boolean switchObject(SlitherLinkMarkerOptions markerOption, SlitherLinkGameMove move) {
        F<SlitherLinkObject, SlitherLinkObject> f = obj -> {
            switch (obj) {
            case Empty:
                return markerOption == SlitherLinkMarkerOptions.MarkerBeforeLine ?
                        SlitherLinkObject.Marker : SlitherLinkObject.Line;
            case Line:
                return markerOption == SlitherLinkMarkerOptions.MarkerAfterLine ?
                        SlitherLinkObject.Marker : SlitherLinkObject.Empty;
            case Marker:
                return markerOption == SlitherLinkMarkerOptions.MarkerBeforeLine ?
                        SlitherLinkObject.Line : SlitherLinkObject.Empty;
            }
            return obj;
        };
        SlitherLinkObject[] dotObj = get(move.p);
        move.obj = f.f(dotObj[move.objOrientation == SlitherLinkObjectOrientation.Horizontal ? 1 : 2]);
        return setObject(move);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/slitherlink/SlitherLinkGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.slitherlink

import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState
import com.zwstudio.logicpuzzlesandroid.common.domain.GameOperationType
import com.zwstudio.logicpuzzlesandroid.common.domain.GridLineObject
import com.zwstudio.logicpuzzlesandroid.common.domain.HintState
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class SlitherLinkGameState(game: SlitherLinkGame) : CellsGameState<SlitherLinkGame, SlitherLinkGameMove, SlitherLinkGameState>(game) {
    val objArray = Array(rows * cols) { Array(4) { GridLineObject.Empty } }
    val pos2state = mutableMapOf<Position, HintState>()

    operator fun get(row: Int, col: Int) = objArray[row * cols + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, dotObj: Array<GridLineObject>) {objArray[row * cols + col] = dotObj}
    operator fun set(p: Position, obj: Array<GridLineObject>) {this[p.row, p.col] = obj}
    private fun isValidMove(move: SlitherLinkGameMove) = !(move.p.row == rows - 1 && move.dir == 2 || move.p.col == cols - 1 && move.dir == 1)

    init {
        updateIsSolved()
    }

    override fun setObject(move: SlitherLinkGameMove): GameOperationType {
        if (!isValidMove(move)) return GameOperationType.Invalid
        val dir = move.dir
        val dir2 = (dir + 2) % 4
        val p1 = move.p
        val p2 = p1 + SlitherLinkGame.offset[dir]
        val o = this[p1][dir]
        if (o == move.obj) return GameOperationType.Invalid
        this[p1][dir] = move.obj
        this[p2][dir2] = this[p1][dir]
        updateIsSolved()
        return GameOperationType.MoveComplete
    }

    override fun switchObject(move: SlitherLinkGameMove): GameOperationType {
        if (!isValidMove(move)) return GameOperationType.Invalid
        val markerOption = MarkerOptions.entries[game.gdi.markerOption]
        move.obj = when (val o = this[move.p][move.dir]) {
            GridLineObject.Empty -> if (markerOption == MarkerOptions.MarkerFirst) GridLineObject.Marker else GridLineObject.Line
            GridLineObject.Line -> if (markerOption == MarkerOptions.MarkerLast) GridLineObject.Marker else GridLineObject.Empty
            GridLineObject.Marker -> if (markerOption == MarkerOptions.MarkerFirst) GridLineObject.Line else GridLineObject.Empty
            else -> o
        }
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 3/SlitherLink

        Summary
        Draw a loop a-la-minesweeper!

        Description
        1. Draw a single looping path with the aid of the numbered hints. The path
           cannot have branches or cross itself.
        2. Each number in a tile tells you on how many of its four sides are touched
           by the path.
        3. For example:
        4. A 0 tells you that the path doesn't touch that square at all.
        5. A 1 tells you that the path touches that square ONLY one-side.
        6. A 3 tells you that the path does a U-turn around that square.
        7. There can't be tiles marked with 4 because that would form a single
           closed loop in it.
        8. Empty tiles can have any number of sides touched by that path.
    */
    private fun updateIsSolved() {
        isSolved = true
        // 2. Each number in a tile tells you on how many of its four sides are touched
        // by the path.
        for ((p, n2) in game.pos2hint) {
            val n1 = (0..<4).count { i -> this[p + SlitherLinkGame.offset2[i]][SlitherLinkGame.dirs[i]] == GridLineObject.Line }
            pos2state[p] = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            if (n1 != n2) isSolved = false
        }
        if (!isSolved) return
        val pos2dirs = mutableMapOf<Position, List<Int>>()
        for (r in 0..<rows)
            for (c in 0..<cols) {
                val p = Position(r, c)
                val dirs = (0..<4).filter { this[p][it] == GridLineObject.Line }
                if (dirs.size == 2)
                    // 1. Draw a single looping path
                    pos2dirs[p] = dirs
                else if (dirs.isNotEmpty()) {
                    // 1. The path cannot have branches or cross itself.
                    isSolved = false; return
                }
            }
        // Check the loop
        val p = pos2dirs.keys.firstOrNull()
        if (p == null) { isSolved = false; return }
        var p2 = p
        var n = -1
        while (true) {
            val dirs = pos2dirs[p2]
            if (dirs == null) { isSolved = false; return }
            pos2dirs.remove(p2)
            n = dirs.first { (it + 2) % 4 != n }
            p2 += SlitherLinkGame.offset[n]
            if (p2 == p) break
        }
    }
}